### PR TITLE
Update environments, configurations and scripts.

### LINKED ISSUE #28: Update environments, configurations and scripts.
Update environments, configurations and scripts.

### PR DESCRIPTION
**Pull request by diegovalenzuelaiturra**
Closes https://github.com/TeselaGen/api-client/issues/28

# Contributions

- flatten the repository

- Updated dependencies
   
    - `pyproject.toml` 
        - Added development dependencies
        - Updated some requirements
    
    - `poetry.lock`
        - updated state

- Added `.vscode` folder with development configurations

- Added configuration files for formatters, linters, fixers, pytest, etc

- Add formatters/fixers on `pre-commit`.

- Updated environment definitions and scripts

    - `Dockerfile`
        - updated `poetry` version
        - added `locale` configurations
        - added `git` install
        - joined some `RUN` commands
        - added few sanity checks
        - updated documentation/comments
        - some refactoring

    - `build.sh`
        - removed unnecessary comments
        - some refactoring

    - `run.sh`
        - Added `--init` flag to `docker run` command
        - some refactoring

    - `run_dev.sh`
        - Added `--init` flag to `docker run` command
        - some refactoring

- Applied end-of-line normalization
    - Mostly normalized `.*csv` files and few `*.py` files

- Applied formatters and some minor refactoring actions

- Fixed few bugs

# Tests

Currently, the following tests are failing when tested against `platform`

Library:

- [x] `teselagen/api/tests/test_discover_client.py::TestDISCOVERClient::test_get_models_by_type[evolutive]`
    - Fails with the following error
      ```diff
      - OSError: Can't found 'data' key in response: {'message': 'Submission success.'}
      ``` 

- [x] `teselagen/api/tests/test_discover_client.py::TestDISCOVERClient::test_get_model_submit_get_cancel_delete`
    - Fails with the following error
      ```diff
      - Exception: Got code : 500. Reason : Internal Server Error
      ``` 

- [x] `teselagen/api/tests/test_multiomics.py::TestTESTClientMultiomicsData::test_download_file`

- [x] `teselagen/api/tests/test_test_client.py::TestTESTClient::test_download_file`

Notebooks:

- [x] `teselagen/examples/pytested/Hello-World-TEST-module.ipynb::nbregression(Hello-World-TEST-module)`
    - Related to `download_file` method
    - Fails with the following error
      ```diff
      - Exception: Got code : 500. Reason : Internal Server Error
      ``` 
- [x] `teselagen/examples/pytested/Closing-the-DBTL-Cycle.ipynb::nbregression(Closing-the-DBTL-Cycle)`
    - Related to `get_models_by_type` method 
    - Fails with the following error
      ```diff
      - OSError: Can't found 'data' key in response: {'message': 'Submission success.'}
      ``` 

# Update 1

- Clean up is performed automatically before and after the tests are run, by defining `pytest_sessionstart` and `pytest_sessionfinish` PyTest hooks in the the `conftest.py`

- Fixed few tests by renaming files, column names, and by fixing a URL

# Notes

Few files were moved, but they are displayed as if they were deleted and then created. But history is being preserved and can be explored with `git log --follow <filepath>` ( https://stackoverflow.com/a/2314757/15999297 )

```bash
# Example
git log --follow poetry.lock
```

### CODE REVIEW

File: teselagen/api/tests/test_multiomics.py [hunk 1]
```diff
@@ -347,82 +492,137 @@ def upload_transcriptomics( temp_dir, test_data, client_with_lab, wild_type_expe
...
+        test_data: Dict[str, pd.DataFrame],
+        host_url: str,
+    ):
+        """Check files download."""
+
+        # if "platform.teselagen.com" in host_url:
+        #     # NOTE: No other code is executed after the pytest.xfail() call, differently from the pytest.mark.xfail()
+        #     pytest.xfail(reason=("These endpoints are under maintenance on the platform. "
+        #                          "This should be solved in the following updates. "
+        #                          "Please, contact the TeselaGen team for more information. "))
```
Comment (juanneilson) [inline teselagen/api/tests/test_multiomics.py:625]: Is this currently valid?

**Review by juanneilson — COMMENTED** [review]
We should consider looking into notebooks tests that fail against `platform`:
```
FAILED teselagen/examples/pytested/Closing-the-DBTL-Cycle.ipynb::nbregression(Closing-the-DBTL-Cycle)
FAILED teselagen/examples/pytested/Hello-World-TEST-module.ipynb::nbregression(Hello-World-TEST-module)
```

**Comment by juanneilson** [comment]
Check why are failing the notebooks (against `platform` and `RC`) and `get_models_by_type` (against RC)

**Comment by diegovalenzuelaiturra** [comment]
.

**Comment by diegovalenzuelaiturra** [comment]
Cleaning up some test `files` and `assays` before the test run starts fixes most of the tests that were broken - The clean up has been implemented as a _pytest hook_ at `pytest_sessionstart` (and also at `pytest_sessionfinish`)

- [x] `teselagen/api/tests/test_discover_client.py::TestDISCOVERClient::test_get_models_by_type[evolutive]`

- [x] `teselagen/api/tests/test_test_client.py::TestTESTClient::test_download_file`

- [x] `teselagen/api/tests/test_discover_client.py::TestDISCOVERClient::test_get_model_submit_get_cancel_delete`

- [x] `teselagen/api/tests/test_multiomics.py::TestTESTClientMultiomicsData::test_optical_density_upload`


- [x] `teselagen/api/tests/test_multiomics.py::TestTESTClientMultiomicsData::test_upload_external_metabolites`


- [x] `teselagen/api/tests/test_multiomics.py::TestTESTClientMultiomicsData::test_download_file`

The following test has been fixed by re-creating the seed data/models
- [x] `teselagen/examples/pytested/Closing-the-DBTL-Cycle.ipynb::nbregression(Closing-the-DBTL-Cycle)`
    - `OSError: Didn't found model`

File: teselagen/examples/pytested/Hello-World-TEST-module.ipynb [hunk 2]
```diff
@@ -562,10 +563,10 @@
     }
    ],
    "source": [
-    "## Upload a File with an Assay.\n",
+    "# Upload a File with an Assay.\n",
     "filepath = str(get_project_root() / \"teselagen/examples/test_module_example.csv\")\n",
```
Comment (juanneilson) [inline teselagen/examples/pytested/Hello-World-TEST-module.ipynb]:
```suggestion
    "filepath = str(get_project_root() / \"teselagen/api/tests/example_file.csv\")\n",
```

**Review by juanneilson — CHANGES_REQUESTED** [review]

File: teselagen/examples/pytested/Hello-World-TEST-module.ipynb (hunk 2 above)
Comment (diegovalenzuelaiturra) [reply teselagen/examples/pytested/Hello-World-TEST-module.ipynb]: done

**Review by juanneilson — APPROVED** [review]

**Review by AndresPerezTesela — APPROVED** [review]

**Pull request merged**